modules/controller/index.js

```diff
@@ -1,193 +1,31 @@
 const _ = require('lodash');
 
-const sourceHas = require('./lib/source_has');
-const parseOptions = require('./lib/parse_options');
 
-const payloads = {
-  create: require('./lib/payloads/create'),
-  read: require('./lib/payloads/read'),
-  readRelation: require('./lib/payloads/read'),
-  update: require('./lib/payloads/update'),
-  destroy: require('./lib/payloads/destroy')
-};
+const configureController = require('./lib/configure_controller');
+const validateController = require('./lib/validate_controller');
+const requestHandler = require('./lib/request_handler');
 
-function Controller(opts) {
-  _.extend(this, parseOptions(opts));
-}
-
-Controller.prototype._extract = require('./lib/extract');
-Controller.prototype._requestHandler = require('./lib/request_handler');
-Controller.prototype.responder = require('./lib/responder');
-
-Controller.prototype._validateController = function (config) {
-  var source = this.source;
-  var method = config.sourceMethod;
+const METHODS = ['create', 'read', 'readRelation', 'update', 'destroy'];
 
-  return _.compose(_.flatten, _.compact)([
-    sourceHas(source.relations(), config.include, 'relations'),
-    sourceHas(source.filters(), Object.keys(config.filters), 'filters'),
-    // this is crap
-    (method === 'read' || method === 'readRelation') ? null :
-      sourceHas(
-        method === 'create' ? source.model : source.model.prototype,
-        config.method,
-        'method'
-      )
-  ]);
-};
-
-Controller.prototype._throwIfModel = function(model) {
-  if (model) {
-    var err = new Error('Model with this ID already exists');
-    err.httpStatus = 409;
-    err.title = 'Conflict';
-    throw err;
+function Controller(opts) {
+  if (!opts) {
+    opts = {};
   }
-};
-
-Controller.prototype._throwIfNoModel = function(model) {
-  if (!model) {
-    var err = new Error('Unable to locate model.');
-    err.httpStatus = 404;
-    err.title = 'Not found';
-    throw err;
+  if (!opts.source) {
+    throw new Error('No source specified.');
   }
+  _.extend(this, opts);
+}
 
-  // Bookshelf throws an error for any number of unrelated reasons.
-  // json-api requires we throw specific errors for certain situations.
-  if (model instanceof Error) {
-    if (/No rows were affected/.test(model.message)) {
-      model.httpStatus = 404;
-      model.title = 'Not found';
-    } else {
-      model.httpStatus = 500;
-      model.title = 'Server Error';
+METHODS.forEach(function (method) {
```

Comment:
Do you think this is an abstraction too far?  I sort of do, but they're literally all the same now, save for the method name.